### ISSUE TITLE
Rerank / Cross Endode first vector search

### ISSUE THREAD

**Issue by joonaskit** [joonaskit/Summa#13]
# Rerank / Cross-Encoder Post-Processing for Vector Search

## Summary

`RagService._vector_search()` currently uses **bi-encoder (embedding) similarity search** via ChromaDB `similarity_search()`. Bi-encoders are fast but imprecise — they rank candidates by cosine distance in a shared embedding space, which can surface semantically adjacent but contextually irrelevant chunks. Adding a **cross-encoder reranker** as a second-pass step would significantly improve answer quality by re-scoring the initial candidate set with a model that jointly encodes query and passage together.

This issue covers the research into suitable approaches and, if applicable, an implementation plan.

---

## Background — Bi-Encoder vs Cross-Encoder

| | Bi-Encoder | Cross-Encoder |
|---|---|---|
| **How it works** | Query and passage are encoded independently; similarity is a dot/cosine product | Query and passage are concatenated and passed through the model jointly |
| **Speed** | Very fast (embedding is pre-computed) | Slow per-pair (must run inference on every candidate) |
| **Accuracy** | Good recall, moderate precision | High precision |
| **Typical use** | First-stage retrieval (ANN search) | Second-stage reranking of top-k candidates |

The standard two-stage RAG pattern:
1. **Retrieve** a larger candidate set (e.g. `k=20`) with the fast bi-encoder.
2. **Rerank** that set with a cross-encoder and take the top-`n` (e.g. `n=4`) for context assembly.

---

## Current State

- **`RagService._vector_search()`** (`backend/services.py`, line 881) calls `self.vectorstore.similarity_search(query, k=k)` — pure bi-encoder retrieval via ChromaDB.
- **`RagService.query_with_context()`** (line 887) calls `_vector_search(query, k=4)` and sends all 4 results to the LLM without any reranking.
- No reranking or cross-encoder library is currently listed as a dependency.

---

### Option A — `sentence-transformers` CrossEncoder (local, free)

The [`sentence-transformers`](https://www.sbert.net/docs/cross_encoder/usage/usage.html) library provides a `CrossEncoder` class that wraps HuggingFace models.

**Recommended models:**

| Model | Size | Notes |
|---|---|---|
| `cross-encoder/ms-marco-MiniLM-L-6-v2` | ~67 MB | Fast, English, strong MS-MARCO performance |
| `cross-encoder/ms-marco-MiniLM-L-12-v2` | ~134 MB | More accurate, slightly slower |
| `BAAI/bge-reranker-base` | ~280 MB | Multilingual-friendly, strong general purpose |
| `BAAI/bge-reranker-large` | ~560 MB | Best quality, higher cost |

**Usage sketch:**
```python
from sentence_transformers import CrossEncoder

reranker = CrossEncoder("cross-encoder/ms-marco-MiniLM-L-6-v2")
pairs = [(query, doc.page_content) for doc in candidates]
scores = reranker.predict(pairs)
reranked = sorted(zip(scores, candidates), key=lambda x: x[0], reverse=True)
top_docs = [doc for _, doc in reranked[:n]]
```

**Pros:** Free, local, no extra API, works off-the-shelf with existing stack.  
**Cons:** Adds a dependency (`sentence-transformers`, `torch`); inference adds ~100–500 ms per query depending on hardware and model size; model must be downloaded on first use.

---

### Option B — LM Studio / OpenAI-Compatible Reranker

Some LM Studio deployments expose a `/v1/rerank` endpoint (Jina, Cohere-compatible). If the local LM Studio instance supports this it could be called via a simple HTTP request — no new Python library needed.

**Pros:** Consistent with the existing `base_url` pattern already used in `LLMService` and `RagService`.  
**Cons:** LM Studio rerank support is limited (model-dependent); adds a network round-trip; harder to guarantee availability.

---

### Option C — Cohere Rerank API (cloud)

Cohere offers a managed `/v1/rerank` endpoint (`cohere` Python SDK or raw HTTP).

**Pros:** State-of-the-art quality, simple API, no GPU needed.  
**Cons:** Requires internet + API key; costs money; against the offline-first philosophy of Summa.

---

### Option D — LangChain `ContextualCompressionRetriever` + `CrossEncoderReranker`

LangChain provides a built-in integration:

```python
from langchain.retrievers import ContextualCompressionRetriever
from langchain.retrievers.document_compressors import CrossEncoderReranker
from langchain_community.cross_encoders import HuggingFaceCrossEncoder

reranker_model = HuggingFaceCrossEncoder(model_name="BAAI/bge-reranker-base")
compressor = CrossEncoderReranker(model=reranker_model, top_n=4)
compression_retriever = ContextualCompressionRetriever(
    base_compressor=compressor,
    base_retriever=self.vectorstore.as_retriever(search_kwargs={"k": 20})
)
results = compression_retriever.invoke(query)
```

**Pros:** Integrates cleanly with the existing LangChain stack; declarative; easy to swap reranker.  
**Cons:** Still requires `sentence-transformers`; introduces a new LangChain retriever abstraction on top of the existing code.

---

## Recommended Approach

**Option A or D** (both local, free, no API key) with `cross-encoder/ms-marco-MiniLM-L-6-v2` as the default model for its small footprint and strong English performance. Option D (LangChain native) is preferred if we want to stay consistent with the LangChain abstraction layer already in place.

The reranker should be:
- **Optional / feature-flagged** (env var `ENABLE_RERANK=true`) to avoid forced inference overhead for users without a GPU.
- **Configurable** — model name via env var `RERANK_MODEL`.
- Applied only to the persistent ChromaDB path (not the in-memory store used for "Chat with a file").

---

### 1. `backend/services.py` — `RagService`

- Add an optional `CrossEncoder` (or `ContextualCompressionRetriever`) initialized in `__init__` when `ENABLE_RERANK=true`.
- Modify `_vector_search()` (or add `_rerank()`) to apply reranking on the initial `k * rerank_factor` candidates, then return the top `k`.

```python
# Pseudocode — exact implementation TBD
def _vector_search(self, query: str, k: int = 4):
    fetch_k = k * self.rerank_factor if self.reranker else k
    candidates = self.vectorstore.similarity_search(query, k=fetch_k)
    if self.reranker:
        candidates = self._rerank(query, candidates, top_n=k)
    return candidates

def _rerank(self, query: str, docs: list, top_n: int):
    pairs = [(query, d.page_content) for d in docs]
    scores = self.reranker.predict(pairs)
    ranked = sorted(zip(scores, docs), key=lambda x: x[0], reverse=True)
    return [doc for _, doc in ranked[:top_n]]
```

### 2. `backend/requirements.txt` / `Dockerfile`

- Add `sentence-transformers` (and `torch` if not already present) under an optional extras section or behind a build arg.

### 3. Environment / Config

| Variable | Default | Description |
|---|---|---|
| `ENABLE_RERANK` | `false` | Enable cross-encoder reranking |
| `RERANK_MODEL` | `cross-encoder/ms-marco-MiniLM-L-6-v2` | HuggingFace model ID |
| `RERANK_FETCH_K` | `20` | Number of candidates to fetch before reranking |

---

## Acceptance Criteria

- [ ] Research complete — approach selected and documented.
- [ ] `RagService` can be initialized with reranking enabled via env var.
- [ ] When enabled, `_vector_search` fetches `RERANK_FETCH_K` candidates and returns top-`k` after reranking.
- [ ] When disabled, behaviour is identical to current implementation.
- [ ] Reranker model is downloaded on first run and cached; subsequent starts are fast.
- [ ] Answer quality is subjectively improved on multi-chunk queries (manual spot-check).
- [ ] No performance regression when reranking is disabled.

---

## Files to Change

| File | Change |
|---|---|
| `backend/services.py` | Add reranker init + `_rerank()` method, modify `_vector_search()` |
| `backend/requirements.txt` | Add `sentence-transformers` |
| `backend/Dockerfile` (if exists) | Ensure `torch` / `sentence-transformers` are installed |
| `.env.example` | Document `ENABLE_RERANK`, `RERANK_MODEL`, `RERANK_FETCH_K` |

---

## Out of Scope

- Switching the primary bi-encoder or embedding model.
- Training or fine-tuning a custom reranker on domain data.
- Exposing reranking scores in the API response or UI.
- Applying reranking to the in-memory vector store (used for "Chat with a file").

---

## References

- [SBERT Cross-Encoder Docs](https://www.sbert.net/docs/cross_encoder/usage/usage.html)
- [LangChain CrossEncoderReranker](https://python.langchain.com/docs/integrations/document_transformers/cross_encoder_reranker/)
- [BAAI/bge-reranker-base on HuggingFace](https://huggingface.co/BAAI/bge-reranker-base)
- [Arxiv — Passage Re-ranking with BERT](https://arxiv.org/abs/1901.04085)
